docs/architecture/adr/ADR040-cluster-stability-node-replacement.md

```diff
@@ -0,0 +1,137 @@
+
+SREs sometimes need to change various things about a cluster's worker nodes, and therefore replace all of the nodes:
+* AMI (e.g. for an EKS update)
+* instance type
+* anything else in the launch template in some way (e.g. instance role)
+
+In addition to this, in our Terraform we get the latest AMI ID for our EKS version from AWS - this means that as soon as AWS releases a new AMI for our current EKS version, the next time the cluster's deployer step runs for Terraform it will also replace all the nodes.
+
+We have a CloudFormation UpdatePolicy on our worker node auto-scaling-groups that tells it to replace 1 node at a time in each ASG. This would be fine if we didn't have 3 ASGs with a small number of nodes in each - right now this practice terminates too many instances at once and results in the cluster becoming unstable. This in turn causes outages in the applications running in the cluster.
+
+Without such a policy we'd be able to update the launch template for the ASG and it wouldn't remove any existing nodes, just set up any new future nodes correctly.
+
+The reason we have 3 small ASGs is so the cluster auto-scaler can scale nodes independently in each AZ - it will know that a pod attempting (and failing) to schedule is tied to a persistent volume claim in a particular availability zone and can scale up the ASG for that zone.
+
+## Possible actions
+
+Here are some of the actions we may want to consider. It is expected we'll choose some subset/variation of these.
+
+### Action 1
+
+Specify WaitOnResourceSignals, MinSuccessfulInstancesPercent and PauseTime in the CloudFormation UpdatePolicy. Mutually exclusive with action 2. Likely also requires action 4 to be safe.
+
+The [AWS docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html) say this about the WaitOnResourceSignals parameter:
+```
+Specifies whether the Auto Scaling group waits on signals from new instances during an update. Use this property to ensure that instances have completed installing and configuring applications before the Auto Scaling group update proceeds. AWS CloudFormation suspends the update of an Auto Scaling group after new EC2 instances are launched into the group. AWS CloudFormation must receive a signal from each new instance within the specified PauseTime before continuing the update. To signal the Auto Scaling group, use the cfn-signal helper script or SignalResource API.
+```
+
+Sometimes new worker nodes don't appear ready in Kubernetes until after the previous worker node they're replacing has been terminated. This should allow us to resolve that and ensure that for every instance terminated e.g. through launch template changes, we have a replacement already in-place.
+
+We could use one of these approaches:
+* When a worker node starts we could have it schedule a Job, selecting itself to run on, that calls SignalResource.
+* We could have a DaemonSet that calls SignalResource and then sleeps infinitely
```

Comment:
I don't like the idea of keeping a pod running using memory, IP addr, resources ... feels like this should be a Job or maybe even just part of the cloudinit/userdata stuff on boot?

Would also be nice if the logic for waiting for actually "all pods are running ok" rather than just "is up" ... not sure is that was implied or not